docs/pallas/gpu/reference.md

```diff
@@ -89,24 +89,169 @@ TODO
 
 ### Hopper (`wgmma`)
 
-TODO
+In this section, we cover the basics of using the Hopper-generation TensorCores, exposed in
+PTX as the [`wgmma.mma_async` instruction](https://docs.nvidia.com/cuda/parallel-thread-execution/#asynchronous-warpgroup-level-matrix-instructions-wgmma-mma).
+
+#### Allocating the accumulator
+
+In the Hopper hardware architecture the accumulator is allocated in registers, but in Pallas
+it is modeled as a mutable reference, as each MMA operation accumulates in-place.
+There are two ways to allocate the accumulator.
+
+To create a zero-initialized accumulator you can use `pl.run_scoped` with a
+`plgpu.ACC((m, n), dtype)` type.
+```python
+def compute(acc_ref):
+  ...
+  return acc_ref[...]
+output = pl.run_scoped(compute, plgpu.ACC((m, n), jnp.float32))
+```
+Dereferencing the accumulator reference, as seen in the end of the `compute` function will
+implicitly await all outstanding WGMMA operations.
+
+If you'd like to initialize it with an existing array, you can use `pl.run_state` with
+`plgpu.ACC.init(init_array)`:
+```python
+def compute(acc_ref):
+  ...
+  return # pl.run_state only returns the final value of the accumulator
+output = pl.run_state(compute)(plgpu.ACC.init(init_array))
+```
+If `pl.run_state` has accumulator operands, it implicitly awaits all outstanding WGMMA
+operations before returning the final values.
+
+#### Preparing the `A` and `B` operands
+
+As discussed above, we recommend passing in `A` and `B` through shared memory. In this
+case the correct tiling and swizzling transforms must be specified.
+
+`plgpu.wgmma` additionally allows passing in `A` through registers (i.e. not an SMEM
+reference but as a regular JAX array). This mode, however, comes with a number of
+significant drawbacks and it is very difficult to ensure sufficient synchronization to
+make this safe.
+
+TODO: Explain the conditions under which it is acceptable to do this.
+
+#### Issuing the operation
+
+The supported MMA shapes are such that:
+* `M` is divisible by 64
+* `N` is divisible by 8 and smaller than 256
+* `K` is a multiple of `swizzle` divided by the bytewidth of element type
+
+The currently supported data types are: `jnp.float32`, `jnp.bfloat16` and `jnp.float16`.
+The accumulator `D` must be a `jnp.float32`, with the exception of `jnp.float16` inputs,
+in which case it is allowed to be `jnp.float16` as well.
+
+#### Waiting for the operation to complete
+
+Each `plgpu.wgmma` call implicitly synchronizes with all previous `plgpu.wgmma` calls, such
+that once control returns from it, we guarantee that no WGMMA other than the last issued
+one is still running. As such, any SMEM regions that were read by previously issued WGMMA
+instructions can be reused. This is especially relevant for pipelining WGMMA with async memory copies:
+```python
+buffers = 3  # In reality you might want even more
+assert a_smem.shape == (buffers, m, k)
+assert b_smem.shape == (buffers, k, n)
```

Comment:
Should we also add assertions of the transforms here for completeness?

Reply:
The transforms at this point use internal data structures that represent the inverse of what the user specifies, so it might turn out to be more confusing than helpful